### PR TITLE
Added `pkcs5_pbkdf2_hmac` to `lib_crypto.cr` and `pkcs5.cr`

### PR DESCRIPTION
**Pull request by aurimasniekis**
Added `pkcs5_pbkdf2_hmac` function to support pbkdf2 with multiple algorithm

### CODE REVIEW

**Comment by aurimasniekis** [comment]
libcrypto was not found if I understood correctly the CircleCI error...

**Comment by sdogruyol** [comment]
Hey @aurimasniekis, welcome and congrats for your first contribution to Crystal :tada:

**Comment by aurimasniekis** [comment]
Hi, @sdogruyol thanks your welcoming! I can't figure out why build doesn't want to work. I tried the same solution which helped to build on my local machine but it doesn't work. The most interesting thing that original `PKCS5_PBKDF2_HMAC_SHA1` is found but additional `PKCS5_PBKDF2_HMAC` method is not found then it's in the same lib. Even then in OpenSSL `PKCS5_PBKDF2_HMAC_SHA1` wraps a call `PKCS5_PBKDF2_HMAC`

**Comment by Sija** [comment]
@aurimasniekis See https://github.com/samr7/vanitygen/issues/45.

**Comment by aurimasniekis** [comment]
@Sija thanks for the suggestion, I will try it out. But why then previous `PKCS5_PBKDF2_HMAC_SHA1` works? :D

**Comment by Sija** [comment]
@aurimasniekis I guess because it has been changed between OpenSSL versions `0.9` and `1.0`, see [source for 0.9.8](https://github.com/openssl/openssl/blob/OpenSSL_0_9_8-stable/crypto/evp/p5_crpt2.c#L82) to see what I mean.

**Comment by Sija** [comment]
@aurimasniekis FYI, `PKCS5_PBKDF2_HMAC ` was added in https://github.com/openssl/openssl/commit/856640b54f3d22a34d6565e9c78c441a15222577

**Comment by aurimasniekis** [comment]
I modified `Makefile` to accept custom ENV variable `CRYSTAL_FLAGS` to pass it to the `crystal build` so that I can pass in `bin/ci` flag `--link-flags -L/usr/local/opt/openssl/lib` and then everything works.

File: Makefile [hunk 1]
```diff
@@ -27,7 +27,7 @@ static ?=       ## Enable static linking
 O := .build
 SOURCES := $(shell find src -name '*.cr')
 SPEC_SOURCES := $(shell find spec -name '*.cr')
-FLAGS := $(if $(release),--release )$(if $(stats),--stats )$(if $(progress),--progress )$(if $(threads),--threads $(threads) )$(if $(debug),-d )$(if $(static),--static )
+FLAGS := $(if $(release),--release )$(if $(stats),--stats )$(if $(progress),--progress )$(if $(threads),--threads $(threads) )$(if $(debug),-d )$(if $(static),--static )$(if $(CRYSTAL_FLAGS),$(CRYSTAL_FLAGS) )
```
Comment (Sija) [inline Makefile]: Splitting `$(CRYSTAL_FLAGS)` changes into a separate line for sake of clarity might be a good thing here, WDYT @aurimasniekis?

File: spec/std/openssl/pkcs5_spec.cr [hunk 2]
```diff
@@ -12,4 +12,15 @@ describe OpenSSL::PKCS5 do
...
   end
+
+  it "computes pbkdf2_hmac" do
+    [
+      {1, 16, "0c60c80f961f0e71f3a9b524af601206"},
+      {1, 32, "0c60c80f961f0e71f3a9b524af6012062fe037a6e0f0eb94fe8fc46bdc637164"},
+      {2**16, 16, "1b345dd55f62a35aecdb9229bc7ae95b"},
+      {2**16, 32, "1b345dd55f62a35aecdb9229bc7ae95b305a8d538940134627e46f82d3a41e5e"},
+    ].each do |(iterations, key_size, expected)|
+      OpenSSL::PKCS5.pbkdf2_hmac(:sha1, "password", "salt", iterations, key_size).hexstring.should eq expected
```
Comment (Sija) [inline spec/std/openssl/pkcs5_spec.cr]: Could you add samples for other algos as well?

File: Makefile (hunk 1 above)
Comment (aurimasniekis) [reply Makefile]: If you think it's better than yes I could do it. But personally, I do not think many people go into Makefile

File: spec/std/openssl/pkcs5_spec.cr (hunk 2 above)
Comment (aurimasniekis) [reply spec/std/openssl/pkcs5_spec.cr]: Yes, I should do it... I was bit lazy...

File: Makefile (hunk 1 above)
Comment (Sija) [reply Makefile]: I think so, yes. ATM it's the longest line in the whole file and chis change would make it even longer, no nice.

File: spec/std/openssl/pkcs5_spec.cr [hunk 3]
```diff
@@ -12,4 +12,15 @@ describe OpenSSL::PKCS5 do
       OpenSSL::PKCS5.pbkdf2_hmac_sha1("password", "salt", iterations, key_size).hexstring.should eq expected
     end
   end
+
+  it "computes pbkdf2_hmac" do
+    [
```
Comment (bew) [inline spec/std/openssl/pkcs5_spec.cr]: You could use a tuple here instead of an array

**Comment by Sija** [comment]
@aurimasniekis 🏓

**Comment by ysbaddaden** [comment]
We don't have a strong policy about supported OpenSSL versions in Crystal, but v0.9.8 is still used and deployed a lot.

If an OpenSSL feature is only available on "recent" OpenSSL release, there should be an OpenSSL version check to enable the feature, just like we do for other OpenSSL features that are only available in v1.0 or v1.1.

**Comment by aurimasniekis** [comment]
@Sija sorry, I forgot about this PR after holidays
@ysbaddaden Do you by wrapping with version check would be sufficient?

File: spec/std/openssl/pkcs5_spec.cr (hunk 3 above)
Comment (aurimasniekis) [reply spec/std/openssl/pkcs5_spec.cr]: I will try because I just copied from original test case

**Comment by aurimasniekis** [comment]
Could someone explain the CircleCI error
```
Error: formatting 'spec/std/openssl/pkcs5_spec.cr' produced changes
```

**Comment by Sija** [comment]
@aurimasniekis it's a check done by `crystal tool format --check`. always run `crystal tool format` before pushing to avoid failing CI because of it.

**Comment by aurimasniekis** [comment]
@Sija thanks, fixed

**Comment by Willamin** [comment]
Is there any more movement on this PR?

**Comment by kingsleyh** [comment]
Hi - I really need this - I've taken the code and patched my library to get it to work - but it would be really awesome if would be integrated into Crystal so I could avoid the patching

**Comment by straight-shoota** [comment]
Resolved by #6975

**Pull request closed without merging**